### PR TITLE
Cleanup of various virtual destructors

### PR DESCRIPTION
**Pull request by MaxKellermann**
Many virtual destructors are never used as such (while adding overhead), and this PR removes lots of them (plus related cleanups).

### CODE REVIEW

**Comment by MaxKellermann** [comment]
There are compiler warnings:
```
delete called on non-final 'CVideoInfoTag' that has virtual functions but non-virtual destructor
```
I'll add missing the `final` keywords. Stand by.

**Comment by pkerling** [comment]
All classes that are derived from should have either a virtual destructor or a protected non-virtual destructor. (cf https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c35-a-base-class-destructor-should-be-either-public-and-virtual-or-protected-and-nonvirtual)

**Comment by MaxKellermann** [comment]
Not exactly true - this CppCoreGuidelines article isn't phrased properly. The virtual destructor is only needed if `delete` is called on a base class (or as in the example: `std::unique_ptr` is used with a base class, which is technically the same). If only leaf (`final`) classes are ever deleted, the virtual destructor is useless.
But the compiler warnings are about leaf classes which are not declared `final`.

**Comment by MaxKellermann** [comment]
I was wrong in this case - `CVideoInfoTag` is indeed not a leaf class, and there are other classes derived from it. So there needs to be a virtual destructor after all! There wasn't one before - it was implicitly declared by deriving from `IArchivable`, `ISerializable` and `ISortable` which used to have one. So this previously worked only by accident, and I removed that accident.....

**Comment by MaxKellermann** [comment]
Rebased and added explicit (defaulted) virtual destructors to `CVideoInfoTag` and `CPictureInfoTag`.

**Comment by pkerling** [comment]
> Not exactly true - this CppCoreGuidelines article isn't phrased properly. The virtual destructor is only needed if delete is called on a base class

I was thinking about the cases where you removed the virtual destructor, for example `AddonEvents`. IMO you should make the destructor non-virtual and protected in order to prevent accidental deletion via the base class.
Actually I don't really get why `AddonEvents` even is structured like that and why a simple `enum` wouldn't have sufficed for the event type. But out of scope for this PR...

The same is true for all the interface classes you removed the virtual destructors from. You should definitely add a protected non-virtual destructor to prevent misuse.

> So this previously worked only by accident, and I removed that accident.....

The interfaces having virtual destructors was not an accident :-) But it's of course better to be explicit here.

**Comment by MaxKellermann** [comment]
Agree, protected+defaulted destructors are a good safety net in some cases.
I rebased and added a few of those.
The previous Jenkins failures are due to a "new" compiler warning which is not so new:
```/home/jenkins/workspace/FreeBSD/xbmc/pvr/dialogs/GUIDialogPVRTimerSettings.cpp:593:19: warning: unused variable 'now' [-Wunused-variable]```
I have a patch for this, and I'll submit it in a separate PR.

**Comment by MaxKellermann** [comment]
Sigh, had to amend one commit again because one `final` was invalid (occurred only on Android due to the very strange DRM code).

File: xbmc/utils/BooleanLogic.h [hunk 1]
```diff
@@ -86,16 +86,13 @@ class CBooleanLogicOperation : public IXmlDeserializable
   CBooleanLogicValues m_values;
 };
 
-class CBooleanLogic : public IXmlDeserializable
+class CBooleanLogic
```
Comment (pkerling) [inline xbmc/utils/BooleanLogic.h]: Why does this not have any destructor?

File: xbmc/utils/URIUtils.h [hunk 2]
```diff
@@ -28,9 +28,6 @@ class CURL;
 class URIUtils
```
Comment (pkerling) [inline xbmc/utils/URIUtils.h]: Missing `final`?

File: xbmc/utils/BooleanLogic.h (hunk 1 above)
Comment (MaxKellermann) [reply xbmc/utils/BooleanLogic.h]: You're right, this one should have a `protected` defaulted destructor.

File: xbmc/utils/URIUtils.h (hunk 2 above)
Comment (MaxKellermann) [reply xbmc/utils/URIUtils.h]: Hm, better convert to `namespace`. This isn't a class after all - it just abuses the `class` keyword to create a namespaceo.

Comment (pkerling) [reply xbmc/utils/URIUtils.h]: +1

`StringUtils` is similar, but you don't have to change either one in this PR.

**Comment by MaxKellermann** [comment]
Rebased & `protected` destructor added to `CBooleanLogic`.

**Comment by MaxKellermann** [comment]
Jenkins failed:
```
/home/jenkins/workspace/LINUX-64-Wayland/xbmc/utils/ActorProtocol.h: In constructor ‘KODI::WINDOWING::WAYLAND::CWinSystemWayland::CWinSystemWayland()’:
/home/jenkins/workspace/LINUX-64-Wayland/xbmc/utils/ActorProtocol.h:88:3: error: ‘Actor::Protocol::~Protocol()’ is protected
```
So this is where the `protected` destructor approach fails: when somebody uses the base class.
Kodi's approach of deriving from classes for no apparent reasons (in this case `Actor::Protocol` doesn't have any virtual methods), instead of embedding an instance, is questionable. Maybe I should work towards making those "base" classes like `Actor::Protocol` final, so nobody submits code which derives from them...

**Comment by MaxKellermann** [comment]
Reverted `Actor::Protocol::~Protocol()` to `public`. Let's see if this build succeeds.

**Comment by pkerling** [comment]
> So this is where the protected destructor approach fails: when somebody uses the base class.

Yes, that approach is only for classes that are used exclusively to inherit from. Here the only place where it is instantiated directly is in Wayland, so maybe I should change it there.

> in this case Actor::Protocol doesn't have any virtual methods

So? Does that automatically make derivation meaningless?

**Comment by MaxKellermann** [comment]
> So? Does that automatically make derivation meaningless?

No, but derivation is only useful if the derived class is not only technically derived, but also derived by its meaning; if it adds something to the base class.

The design of `Actor::Protocol` is hard to grasp, because it's completely undocumented. There are exactly two code comments, and they explain the obvious, and obvious comments are disposable.
From the name: an instance of this class describes one protocol. But that's wrong. This is not a protocol; this is a generic inter-thread message queue, and a message is an opaque buffer. The protocol is defined by the caller.

So if this is really just a message queue, why would anybody derive from it? Maybe to add functionality. But none of the derived classes add queue-specific functionality. All they do is serve as a namespace containing protocol-specific definitions (opaque to the queue itself!). No derived class accesses these definitions; they are accessed solely from the outside, not even by an instance, but only by the namespace name:
```
          m_controlPort.SendInMessage(COutputControlProtocol::STATS);
```
It does not matter if `m_controlPort` is a `COutputControlProtocol` or a basic `Actor::Protocol` - there isn't even any wrapper ensuring that a `COutputControlProtocol` command can't be sent over a `CMixerControlProtocol` instance. The way this is used here is clumsy at best.
What would have been *somewhat* helpful would have been this choice of syntax:
```
          m_controlPort.SendInMessage(m_controlPort.STATS);
```
This would make it more obvious to the (human) reader, but still the compiler wouldn't care.
So no, in this case, deriving from `Actor::Protocol` doesn't make sense. One could make a wrapper template to ensure type-safety, to validate all protocol commands at compile-time, and that wrapper would embed that `Actor::Protocol` instance instead of deriving from it, providing inline methods wrapping `Actor::Protocol` methods in a type-safe way. Still nobody would derive from the latter.

(This is just my opinion, not the absolute truth, of course - it's a matter of taste and preferred coding style. There are reasons for/against one or another.)

**Comment by MaxKellermann** [comment]
btw. Jenkins failures again not related to this PR.

**Comment by pkerling** [comment]
ping @FernetMenta, Actor is your code so you might want to take a look

@MaxKellermann You probably won't mind if I merge this as-is for the time being?

**Comment by MaxKellermann** [comment]
Of course I don't mind. This PR appears to be ready for merging (no known regression, just a false positive about an older compiler warning), and other design problems are mostly unrelated to this PR ...

**Comment by Rechi** [comment]
> btw. Jenkins failures again not related to this PR.

They are, because for whatever reason the unused warning wasn't reported before.

> You probably won't mind if I merge this as-is for the time being?

Not as long as jenkins reports success.

**Comment by MaxKellermann** [comment]
@Rechi I'll submit a PR for this warning. I have the patch already. No idea why it is regarded new suddenly - it's 2 years old.

**Comment by MaxKellermann** [comment]
Oh you fixed it already.

**Comment by pkerling** [comment]
@Rechi Sorry, was a bit too quick!

But I double-checked that the warning was true already before the PR. So for the record: All failed builds were built correctly and only failed in Jenkins due to one warning marked as new that clang did not flag before for whatever reason (even though it was already true before).

**Pull request merged**